cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -446,24 +478,36 @@ withIndexEntries (RepoIndex repoCtxt repo@RepoSecure{..}) callback =
               return [] -- skip unrecognized file
             mk (_, _fp, Just (Sec.Some (Sec.IndexPkgMetadata _pkgId))) =
               return [] -- skip metadata
-            mk (dirEntry, _fp, Just (Sec.Some (Sec.IndexPkgCabal pkgId))) = do
-              let blockNo = fromIntegral (Sec.directoryEntryBlockNo dirEntry)
-              return [CachePackageId pkgId blockNo]
+            mk (dirEntry, _fp, Just (Sec.Some file@(Sec.IndexPkgCabal pkgId))) = do
+              let blockNo = Sec.directoryEntryBlockNo dirEntry
+              timestamp <- Sec.indexEntryTime `fmap` indexLookupFileEntry dirEntry file
+              return [CachePackageId pkgId blockNo timestamp]
             mk (dirEntry, _fp, Just (Sec.Some file@(Sec.IndexPkgPrefs _pkgName))) = do
+              let blockNo = Sec.directoryEntryBlockNo dirEntry
               content <- Sec.indexEntryContent `fmap` indexLookupFileEntry dirEntry file
-              return $ map CachePreference (parsePreferredVersions content)
-        entriess <- lazySequence $ map mk (Sec.directoryEntries indexDirectory)
+              timestamp <- Sec.indexEntryTime `fmap` indexLookupFileEntry dirEntry file
+              return $ map (\x ->  CachePreference x blockNo timestamp) (parsePreferredVersions content)
+
+        let mk2 :: (Sec.DirectoryEntry, Sec.Some Sec.IndexEntry)
+                -> (Sec.DirectoryEntry, Sec.IndexPath, Maybe (Sec.Some Sec.IndexFile))
+            mk2 (dent, Sec.Some sie) =
+                (dent, Sec.indexEntryPath sie, fmap Sec.Some (Sec.indexEntryPathParsed sie))
+
+        -- dirIdxEnts :: [(Sec.DirectoryEntry, Sec.Some Sec.IndexEntry)]
+        dirIdxEnts <- lazyUnfold indexLookupEntry (Sec.directoryFirst indexDirectory)
+        entriess <- lazySequence $ map (mk . mk2) dirIdxEnts
```

Comment:
Ok, so originally with using the `Sec.directoryEntries` that gave us only the final version of each file, ie not all intermediate revisions. And previously our strategy was to go through the final versions of each file, in file order, and lookup just the ones we're interested in (which in practice is 99% of them).

Now for the new cache we want to go through all revisions, which means all entries in file order. So instead of using `Sec.directoryEntries` which reads from the tar index, we go straight for `Sec.directoryFirst` which is block 0 and iterate through, using `lazyUnfold`.

But we can now significantly simplify this and do it more efficiently. Note that `indexLookupEntry` and `indexLookupFileEntry` are expensive operations that seek in the tar file and read the tar entry at that point. So lets do it exactly once per entry. The current code does it once in the `lazyUnfold indexLookupEntry` and then again in `mk`. But the old `mk` only did that because it had not previously looked up the entry.

So get rid of `mk2`, change `mk` to:

    toCacheEntries :: Sec.DirectoryEntry -> Sec.Some Sec.IndexEntry  -> [IndexCacheEntry]

Yes, pure now because it has already read the `IndexEntry` from disk so everything is available.

Then it can be:

    withIndexEntries (RepoIndex repoCtxt repo@RepoSecure{..}) callback =
        repoContextWithSecureRepo repoCtxt repo $ \repoSecure ->
          Sec.withIndex repoSecure $ \Sec.IndexCallbacks{..} -> do
          -- Incrementally (lazily) read all the entries in the tar file in order,
          -- including all revisions, not just the last revision of each file
          indexEntries <- lazyUnfold indexLookupEntry (Sec.directoryFirst indexDirectory)
          callback [ cacheEntry
                   | (dirEntry, indexEntry) <- indexEntries
                   , cacheEntry <- toCacheEntries dirEntry indexEntry ]
      where
        toCacheEntries :: Sec.DirectoryEntry -> Sec.Some Sec.IndexEntry  -> [IndexCacheEntry]